File: lib/treeFold.hs
data Tree a = Empty
            | Node (Tree a) a (Tree a)
    deriving (Show, Eq)

leaf :: a -> Tree a
leaf x = Node Empty x Empty

treeSize :: Tree a -> Integer
treeSize Empty = 0
treeSize Node (l _ r) = 1 + treeSize l + treeSize r

treeSum :: Tree Integer -> Integer
treeSum Empty = 0
treeSum (Node l x r) = x + treeSum l + treeSum r

treeDepth :: Tree a -> Integer
treeDepth Empty = 0
treeDepth (Node l _ r) = 1 + max (treeDepth 1) (treeDepth 2)

flatten :: Tree a -> [a]
flatten Empty = []
flatten (Node l x r) = flatten l ++ "x" ++ flatten r

-- lets write a treeFold function
treeFold b -> (b -> a -> b -> b) -> Tree a -> b
treeFold e _ Empty = e
treeFold e f (Node l x r) = f (treeFold e f l ) x (treeFold e f r)

treeSize' :: Tree a -> Integer
treeSize' = treeFold(0, (\l _ r -> 1 + l + r))
-- note that treeFold takes three arguments in the signature
-- here we apply two arguments and these are curried

treeSum' :: Tree Integer -> Integer
treeSum' = treeFold(0, (\l x r -> l + x + r))

treeDepth' :: Tree a -> Integer
treeDepth' = treeFold(0, (\l _ r -> 1 + max l r))

flatten' :: Tree a -> [a]
flatten' = treeFold([], (\l x r -> l ++ [x] ++ r))jkjk

-- sources:
-- http://www.seas.upenn.edu/~cis194/spring13/lectures/07-folds-monoids.html

File: lib/treeFold.js
class Tree {}
class TreeEmpty extends Tree{}
class TreeNode extends Tree {
  constructor(value) {
    super();
    this.value = value;
    this.left = new TreeEmpty;
    this.right = new TreeEmpty;
  }
}

const treeSize = (tree) => {
  // closest we can get to pattern matching syntax in javascript
  switch(tree.constructor) {
    case TreeEmpty:
      return 0;
    case TreeNode:
      return 1 + treeSize(tree.left) + treeSize(tree.right);
    default:
      return;
  }
};

const treeSum = (tree) => {
  switch(tree.constructor) {
    case TreeEmpty:
      return 0;
    case TreeNode:
      return tree.value + treeSize(tree.left) + treeSize(tree.right);
    default:
      return;
  }
};

const treeDepth = (tree) => {
  switch(tree.constructor) {
    case TreeEmpty:
      return 0;
    case TreeNode:
      return 1 + Math.max(treeSize(tree.left), treeSize(tree.right));
    default:
      return;
  }
};

const treeFlatten = (tree) => {
  switch(tree.constructor) {
    case TreeEmpty:
      return [];
    case TreeNode:
      return treeSize(tree.left).concat([tree.value]).concat(treeSize(tree.right));
    default:
      return;
  }
};

// lets write a treeFold function
const treeFold = (e, fn, tree) => {
  switch(tree.constructor) {
    case TreeEmpty:
      return e;
    case TreeNode:
      return fn(
        treeFold(e, fn, tree.left),
        tree.value,
        treeFold(e, fn, tree.right)
      );
    default:
      return;
  }
};

const treeSize = treeFold.bind(null, 0, (l, x, r) => 1 + l + r);
const treeSum = treeFold.bind(null, 0, (l, x, r) => x + l + r);
const treeDepth = treeFold.bind(null, 0, (l, x, r) => 1 + Math.max(l, r));
const treeFlatten = treeFold.bind(null, [], (l, x, r) => l.concat([x]).concat(r));
